File: examples/ts/Scripts/JoystickModule.ts
/// <reference path="../fairygui.d.ts" />

import { Callable } from "godot";

type MoveChangedListener = (degree: number) => void;
type EndedListener = () => void;

class SimpleSignal<T extends (...args: never[]) => void> {
    private readonly _listeners: Array<T> = [];

    connect(listener: T): void {
        this._listeners.push(listener);
    }

    emit(...args: Parameters<T>): void {
        for (const listener of this._listeners) {
            listener(...args);
        }
    }
}

export default class JoystickModule {
    readonly move_changed = new SimpleSignal<MoveChangedListener>();
    readonly ended = new SimpleSignal<EndedListener>();

    private _button: GButton | null = null;
    private _thumb: GObject | null = null;
    private _touchArea: GObject | null = null;
    private _center: GObject | null = null;
    private _groot: GRoot | null = null;

    private _initX = 0.0;
    private _initY = 0.0;
    private _startStageX = 0.0;
    private _startStageY = 0.0;
    private _lastStageX = 0.0;
    private _lastStageY = 0.0;
    private readonly _radius = 150;
    private _touchId = -1;

    constructor(mainView: GObject) {
        this._groot = GRoot.getInstance();
        this._button = mainView.getChild("joystick") as GButton | null;
        if (this._button != null) {
            this._thumb = this._button.getChild("thumb");
            this._button.changeStateOnClick = false;
        }
        this._touchArea = mainView.getChild("joystick_touch");
        this._center = mainView.getChild("joystick_center");

        if (this._center != null) {
            this._initX = this._center.getX() + this._center.getWidth() / 2.0;
            this._initY = this._center.getY() + this._center.getHeight() / 2.0;
        }

        if (this._touchArea != null) {
            this._touchArea.addEventListener(
                UIEventDispatcher.TOUCHBEGIN,
                Callable.create(this._onTouchBegin.bind(this)),
            );
            this._touchArea.addEventListener(
                UIEventDispatcher.TOUCHMOVE,
                Callable.create(this._onTouchMove.bind(this)),
            );
            this._touchArea.addEventListener(
                UIEventDispatcher.TOUCHEND,
                Callable.create(this._onTouchEnd.bind(this)),
            );
        }
    }

    private _onTouchBegin(): void {
        if (this._touchId !== -1 || this._groot == null || this._button == null || this._thumb == null) {
            return;
        }
        this._touchId = 0;

        const pt = this._groot.globalToLocal(this._groot.getTouchPosition());
        let bx = pt.x;
        let by = pt.y;

        if (bx < 0) {
            bx = 0;
        }
        if (this._touchArea != null && bx > this._touchArea.getWidth()) {
            bx = this._touchArea.getWidth();
        }
        if (by > this._groot.getHeight()) {
            by = this._groot.getHeight();
        } else if (this._touchArea != null && by < this._touchArea.getY()) {
            by = this._touchArea.getY();
        }

        this._lastStageX = bx;
        this._lastStageY = by;
        this._startStageX = bx;
        this._startStageY = by;
        this._button.selected = true;

        if (this._center != null) {
            this._center.setVisible(true);
            this._center.setPosition(
                bx - this._center.getWidth() / 2.0,
                by - this._center.getHeight() / 2.0,
            );
        }
        this._button.setPosition(
            bx - this._button.getWidth() / 2.0,
            by - this._button.getHeight() / 2.0,
        );

        const deltaX = bx - this._initX;
        const deltaY = by - this._initY;
        const degrees = (Math.atan2(deltaY, deltaX) * 180.0) / Math.PI;
        this._thumb.setRotation(degrees + 90);
    }

    private _onTouchMove(): void {
        if (
            this._touchId === -1 ||
            this._groot == null ||
            this._button == null ||
            this._thumb == null
        ) {
            return;
        }

        const pt = this._groot.globalToLocal(this._groot.getTouchPosition());
        const bx = pt.x;
        const by = pt.y;
        const moveX = bx - this._lastStageX;
        const moveY = by - this._lastStageY;

        this._lastStageX = bx;
        this._lastStageY = by;

        let buttonX = this._button.getX() + moveX;
        let buttonY = this._button.getY() + moveY;
        let offsetX = buttonX + this._button.getWidth() / 2.0 - this._startStageX;
        let offsetY = buttonY + this._button.getHeight() / 2.0 - this._startStageY;

        const rad = Math.atan2(offsetY, offsetX);
        const degree = (rad * 180.0) / Math.PI;
        this._thumb.setRotation(degree + 90);

        const maxX = this._radius * Math.cos(rad);
        const maxY = this._radius * Math.sin(rad);
        if (Math.abs(offsetX) > Math.abs(maxX)) {
            offsetX = maxX;
        }
        if (Math.abs(offsetY) > Math.abs(maxY)) {
            offsetY = maxY;
        }

        buttonX = this._startStageX + offsetX;
        buttonY = this._startStageY + offsetY;
        if (buttonX < 0) {
            buttonX = 0;
        }
        if (buttonY > this._groot.getHeight()) {
            buttonY = this._groot.getHeight();
        }

        this._button.setPosition(
            buttonX - this._button.getWidth() / 2.0,
            buttonY - this._button.getHeight() / 2.0,
        );
        this.move_changed.emit(degree);
    }

    private _onTouchEnd(): void {
        if (this._touchId === -1 || this._button == null || this._thumb == null) {
            return;
        }
        this._touchId = -1;
        this._thumb.setRotation(this._thumb.getRotation() + 180);
        if (this._center != null) {
            this._center.setVisible(true);
            this._center.setPosition(
                this._initX - this._center.getWidth() / 2.0,
                this._initY - this._center.getHeight() / 2.0,
            );
        }
        this._button.setPosition(
            this._initX - this._button.getWidth() / 2.0,
            this._initY - this._button.getHeight() / 2.0,
        );
        this._button.selected = false;
        this._thumb.setRotation(0);
        this.ended.emit();
    }
}

File: examples/csharp/Scripts/JoystickModule.cs
using Godot;
using System;

namespace FairyGUI.Examples;

public class JoystickModule
{
    public event Action<float> MoveChanged;
    public event Action Ended;

    private GuiObject _button;
    private GuiObject _thumb;
    private GuiObject _touchArea;
    private GuiObject _center;
    private GRoot _groot;

    private float _initX;
    private float _initY;
    private float _startStageX;
    private float _startStageY;
    private float _lastStageX;
    private float _lastStageY;
    private int _radius = 150;
    private int _touchId = -1;

    public JoystickModule(GuiObject mainView)
    {
        _groot = GRoot.GetInstance();
        _button = mainView.GetChild("joystick");
        _thumb = _button.GetChild("thumb");
        _touchArea = mainView.GetChild("joystick_touch");
        _center = mainView.GetChild("joystick_center");

        if (_button is GButton btn)
            btn.SetChangeStateOnClick(false);

        if (_center != null)
        {
            _initX = _center.GetX() + _center.GetWidth() / 2.0f;
            _initY = _center.GetY() + _center.GetHeight() / 2.0f;
        }

        _touchArea?.AddEventListener(FguiEvent.TouchBegin, Callable.From(OnTouchBegin));
        _touchArea?.AddEventListener(FguiEvent.TouchMove, Callable.From(OnTouchMove));
        _touchArea?.AddEventListener(FguiEvent.TouchEnd, Callable.From(OnTouchEnd));
    }

    private void OnTouchBegin()
    {
        if (_touchId != -1)
            return;
        _touchId = 0;

        var pt = _groot.GlobalToLocal(_groot.GetTouchPosition());
        var bx = pt.X;
        var by = pt.Y;

        if (bx < 0) bx = 0;
        if (bx > _touchArea.GetWidth()) bx = _touchArea.GetWidth();
        if (by > _groot.GetHeight()) by = _groot.GetHeight();
        else if (by < _touchArea.GetY()) by = _touchArea.GetY();

        _lastStageX = bx;
        _lastStageY = by;
        _startStageX = bx;
        _startStageY = by;

        if (_button is GButton btn)
            btn.SetSelected(true);

        _center.SetVisible(true);
        _center.SetPosition(bx - _center.GetWidth() / 2.0f, by - _center.GetHeight() / 2.0f);
        _button.SetPosition(bx - _button.GetWidth() / 2.0f, by - _button.GetHeight() / 2.0f);

        var deltaX = bx - _initX;
        var deltaY = by - _initY;
        var degrees = Mathf.RadToDeg(Mathf.Atan2(deltaY, deltaX));
        _thumb.SetRotation(degrees + 90);
    }

    private void OnTouchMove()
    {
        if (_touchId == -1)
            return;

        var pt = _groot.GlobalToLocal(_groot.GetTouchPosition());
        var bx = pt.X;
        var by = pt.Y;
        var moveX = bx - _lastStageX;
        var moveY = by - _lastStageY;

        _lastStageX = bx;
        _lastStageY = by;

        var buttonX = _button.GetX() + moveX;
        var buttonY = _button.GetY() + moveY;
        var offsetX = buttonX + _button.GetWidth() / 2.0f - _startStageX;
        var offsetY = buttonY + _button.GetHeight() / 2.0f - _startStageY;

        var rad = Mathf.Atan2(offsetY, offsetX);
        var degree = Mathf.RadToDeg(rad);
        _thumb.SetRotation(degree + 90);

        var maxX = _radius * Mathf.Cos(rad);
        var maxY = _radius * Mathf.Sin(rad);
        if (Mathf.Abs(offsetX) > Mathf.Abs(maxX)) offsetX = maxX;
        if (Mathf.Abs(offsetY) > Mathf.Abs(maxY)) offsetY = maxY;

        buttonX = _startStageX + offsetX;
        buttonY = _startStageY + offsetY;
        if (buttonX < 0) buttonX = 0;
        if (buttonY > _groot.GetHeight()) buttonY = _groot.GetHeight();

        _button.SetPosition(buttonX - _button.GetWidth() / 2.0f, buttonY - _button.GetHeight() / 2.0f);
        MoveChanged?.Invoke(degree);
    }

    private void OnTouchEnd()
    {
        if (_touchId == -1)
            return;
        _touchId = -1;

        _thumb.SetRotation(_thumb.GetRotation() + 180);
        _center.SetVisible(true);
        _center.SetPosition(_initX - _center.GetWidth() / 2.0f, _initY - _center.GetHeight() / 2.0f);
        _button.SetPosition(_initX - _button.GetWidth() / 2.0f, _initY - _button.GetHeight() / 2.0f);

        if (_button is GButton btn)
            btn.SetSelected(false);

        _thumb.SetRotation(0);
        Ended?.Invoke();
    }
}
